### PR TITLE
test: raise Rust and binding coverage

### PR DESCRIPTION
**Pull request by willkill07**
#### Overview

Adds the non-CLI test coverage extracted from the combined Sonar and coverage work for release/0.7.

- [x] I confirm this contribution is my own work, or I have the right to submit it under this project's license.
- [x] I searched existing issues and open pull requests, and this does not duplicate existing work.

#### Details

- Expand Rust core and adaptive unit and integration coverage.
- Cover codec, observability, plugin, logging, worker, subscriber, and runtime edge cases.
- Extend FFI, Python-native, Node, and Go binding coverage while preserving existing test patterns.
- Refactor high-complexity test functions needed by the cognitive-complexity policy PR.
- No breaking API changes.

Validation performed:

- Branch commit hooks, including formatting, Clippy, Cargo checks, Go formatting and vet, and Node formatting
- uv run pre-commit run --all-files on the equivalent combined tree
- Rust workspace tests; the CLI global-state-sensitive suite passed with one test thread
- Exact four-branch merge tree comparison against the original combined branch

#### Where should the reviewer start?

Start with crates/plugin/tests/typed_callbacks.rs for the test-complexity refactors, then crates/core/tests/coverage/logging_rotation_tests.rs for representative new coverage.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- Relates to #632

## Summary by CodeRabbit

* **Tests**
  * Expanded coverage for response caching, logging rotation, codec handling, observability, plugin configuration, worker processes, guardrails, and lifecycle management.
  * Added regression checks for malformed inputs, streaming completion, error handling, status validation, repeated cache updates, and isolated configuration environments.
  * Improved cross-language coverage across FFI, Python, Node.js, Go, and Rust integrations.
  * Strengthened validation of API registries, event metadata, telemetry, callbacks, and stream handling.
  * Refined test organization and shared assertions without changing product behavior.

### CODE REVIEW

File: crates/core/tests/unit/plugins/nemo_guardrails/local_python_tests.rs [hunk 1]
```diff
@@ -531,6 +745,293 @@ fn parse_check_result_rejects_unknown_status() {
...
+fn monitor_test_worker() -> Arc<LocalGuardrailsWorker> {
+    Arc::new(LocalGuardrailsWorker {
+        writer: Mutex::new(None),
+        child: Mutex::new(Command::new("sleep").arg("60").spawn().unwrap()),
+        waiters: Arc::new(Mutex::new(HashMap::new())),
+        stream_events: Arc::new(Mutex::new(HashMap::new())),
+        next_id: AtomicU64::new(0),
+        shutdown_started: AtomicBool::new(false),
+    })
+}
```
[the remark this answers is not in the document]
Comment (willkill07) [reply crates/core/tests/unit/plugins/nemo_guardrails/local_python_tests.rs:958]: No code change needed. LocalGuardrailsWorker implements Drop in python.rs; Drop calls shutdown(), which kills and waits for the child process. Each Arc created by monitor_test_worker is dropped at the end of the helper/test, so the sleep child is reaped rather than orphaned.

File: crates/core/tests/unit/plugins/nemo_guardrails/remote_coverage_tests.rs [hunk 2]
```diff
@@ -205,6 +223,22 @@ fn request_body_and_guardrails_config_helpers_cover_defaults() {
...
+                    "messages": [],
+                    "tools": [{"type": "function", "function": {"name": "lookup"}}]
+                }),
+            },
+            false,
+        ),
+        "does not support OpenAI tool definitions",
+    );
+    runtime.record_access_status(reqwest::StatusCode::OK);
+    runtime.record_access_status(reqwest::StatusCode::OK);
```
[the remark this answers is not in the document]
Comment (willkill07) [reply crates/core/tests/unit/plugins/nemo_guardrails/remote_coverage_tests.rs:242]: Addressed in ccb335bb1. The test now asserts access_state == 2 after both the first successful status and the idempotent second successful status.

File: crates/adaptive/tests/unit/response_cache/intercept_tests.rs [hunk 3]
```diff
@@ -86,3 +86,89 @@ async fn write_behind_returns_eof_before_cache_commit_completes() {
...
+        "incomplete",
+        "in_progress",
+        "queued",
+    ] {
+        assert!(is_error_response(&json!({"status": status})));
+    }
+    assert!(!is_error_response(&json!({"status": "completed"})));
+    assert!(!should_bypass(0.0));
+    assert!(should_bypass(1.0));
+    assert!(next_unit_f64().is_finite());
```
[the remark this answers is not in the document]
Comment (willkill07) [reply crates/adaptive/tests/unit/response_cache/intercept_tests.rs]: Addressed in ccb335bb1. The test now asserts that next_unit_f64 returns a value in [0.0, 1.0).

File: crates/core/tests/unit/observability/plugin_component_tests.rs [hunk 4]
```diff
@@ -3286,3 +3380,136 @@ fn atif_storage_private_helpers_resolve_env_and_key_prefix_branches() {
...
+        .unwrap_err()
+        .to_string()
+        .contains("HTTP ATIF upload")
+    );
+}
+
+#[test]
+#[cfg(feature = "object-store")]
+fn s3_remote_storage_uploads_to_a_custom_http_endpoint() {
+    let _guard = crate::observability::test_mutex().lock().unwrap();
```
[the remark this answers is not in the document]
Comment (willkill07) [reply crates/core/tests/unit/observability/plugin_component_tests.rs]: Addressed in ccb335bb1. The S3 upload test now recovers a poisoned observability mutex using the same pattern as neighboring tests.

File: crates/ffi/tests/integration/plugin_activation_tests.rs [hunk 5]
```diff
@@ -194,6 +160,48 @@ source = "project-file"
...
+
+    // Mutating the file after startup has no effect: discovery is one-shot.
+    std::fs::write(plugins_toml, "invalid = [").expect("mutate plugin config after startup");
+    let intercepted = tool_request_intercepts("ffi-layered-tool", json!({"input": true}));
+    assert_eq!(intercepted["file_static"], true);
+    assert_eq!(intercepted["static_saw_dynamic"], false);
+    assert_eq!(intercepted["native_plugin"], true);
+    assert_eq!(DISCOVERED_STATIC_CALLBACKS.load(Ordering::SeqCst), 1);
+}
+
```
[the remark this answers is not in the document]
Comment (willkill07) [reply crates/ffi/tests/integration/plugin_activation_tests.rs]: Addressed in ccb335bb1. The mutating helper is now named write_and_assert_discovered_activation, and both newly extracted assertion helpers use track_caller.

**Review by ericevans-nv — APPROVED** [review]
